{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "__author__ = 'Jiashun'\n",
    "import re\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "from math import ceil\n",
    "from math import floor\n",
    "\n",
    "# compare single base\n",
    "def SingleBaseCompare(seq1,seq2,i,j):\n",
    "    if seq1[i] == seq2[j]:\n",
    "        return 2\n",
    "    else:\n",
    "        return -1\n",
    "    \n",
    "# Smithâ€“Waterman Alignment \n",
    "def SMalignment(seq1, seq2):\n",
    "    m = len(seq1)\n",
    "    n = len(seq2)\n",
    "    g = -3\n",
    "    matrix = []\n",
    "    for i in range(0, m):\n",
    "        tmp = []\n",
    "        for j in range(0, n):\n",
    "            tmp.append(0)\n",
    "        matrix.append(tmp)\n",
    "    for sii in range(0, m):\n",
    "        matrix[sii][0] = sii*g\n",
    "    for sjj in range(0, n):\n",
    "        matrix[0][sjj] = sjj*g\n",
    "    for siii in range(1, m):\n",
    "        for sjjj in range(1, n):\n",
    "            matrix[siii][sjjj] = max(matrix[siii-1][sjjj] + g, matrix[siii - 1][sjjj - 1] + SingleBaseCompare(seq1,seq2,siii, sjjj), matrix[siii][sjjj-1] + g)\n",
    "    sequ1 = [seq1[m-1]]\n",
    "    sequ2 = [seq2[n-1]]\n",
    "    while m > 1 and n > 1:\n",
    "        if max(matrix[m-1][n-2], matrix[m-2][n-2], matrix[m-2][n-1]) == matrix[m-2][n-2]:\n",
    "            m -= 1\n",
    "            n -= 1\n",
    "            sequ1.append(seq1[m-1])\n",
    "            sequ2.append(seq2[n-1])\n",
    "        elif max(matrix[m-1][n-2], matrix[m-2][n-2], matrix[m-2][n-1]) == matrix[m-1][n-2]:\n",
    "            n -= 1\n",
    "            sequ1.append('-')\n",
    "            sequ2.append(seq2[n-1])\n",
    "        else:\n",
    "            m -= 1\n",
    "            sequ1.append(seq1[m-1])\n",
    "            sequ2.append('-')\n",
    "    sequ1.reverse()\n",
    "    sequ2.reverse()\n",
    "    align_seq1 = ''.join(sequ1)\n",
    "    align_seq2 = ''.join(sequ2)\n",
    "    align_score = 0.\n",
    "    for k in range(0, len(align_seq1)):\n",
    "        if align_seq1[k] == align_seq2[k]:\n",
    "            align_score += 1\n",
    "    align_score = float(align_score)/len(align_seq1)\n",
    "    return align_seq1, align_seq2, align_score\n",
    "\n",
    "# Display BlAST result\n",
    "def Display(seque1, seque2):\n",
    "    le = 60\n",
    "    while len(seque1)-le >= 0:\n",
    "        print('sequence1: ',end='')\n",
    "        for a in list(seque1)[le-40:le]:\n",
    "            print(a,end='')\n",
    "        print(\"\\n\")\n",
    "        print('           ',end='')\n",
    "        for k in range(le-40, le):\n",
    "            if seque1[k] == seque2[k]:\n",
    "                print('|',end='')\n",
    "            else:\n",
    "                print(' ',end='')\n",
    "        print(\"\\n\")\n",
    "        print('sequence2: ',end='')\n",
    "        for b in list(seque2)[le-40:le]:\n",
    "            print(b,end='')\n",
    "        print(\"\\n\")\n",
    "        le += 40\n",
    "    if len(seque1) > le-40:\n",
    "        print('sequence1: ',end='')\n",
    "        for a in list(seque1)[le-40:len(seque1)]:\n",
    "            print(a,end='')\n",
    "        print(\"\\n\")\n",
    "        print('           ',end='')\n",
    "        for k in range(le-40, len(seque1)):\n",
    "            if seque1[k] == seque2[k]:\n",
    "                print('|',end='')\n",
    "            else:\n",
    "                print(' ',end='')\n",
    "        print(\"\\n\")\n",
    "        print('sequence2: ',end='')\n",
    "        for b in list(seque2)[le-40:len(seque2)]:\n",
    "            print(b,end='')\n",
    "        print(\"\\n\")\n",
    "\n",
    "# transform base to numeric value\n",
    "def WordToNum(word):\n",
    "    tmp = []\n",
    "    trans = {'A':1,'C':2,'G':3,'T':4}\n",
    "    for w in word:\n",
    "        tmp.append(trans[w])\n",
    "    return tmp\n",
    "\n",
    "# transform word with 11 bases to its index\n",
    "def WordToIndex(word,word_len):\n",
    "    tmp = 0\n",
    "    word_num = WordToNum(word)\n",
    "    for i,v in enumerate(word_num):\n",
    "        tmp += (v-1)*4**(word_len-i)\n",
    "    return tmp   \n",
    "\n",
    "# Get word's postion in genome from library\n",
    "def GetWordPos(word):\n",
    "    assert len(word)== 11\n",
    "    seek_index = WordToIndex(word,11-1)\n",
    "    positions = []\n",
    "    for chr_name in chr_names:\n",
    "        chr_seq = open('/home/jxiaoae/class/blast/chromosome_{}_library.txt'.format(chr_name),'r')\n",
    "        seeks = np.load(\"chromosome_{}_library_seeks.npy\".format(chr_name))\n",
    "        chr_seq.seek(seeks[seek_index,0])\n",
    "        position = chr_seq.read(seeks[seek_index,1])\n",
    "        try:\n",
    "            positions.append(list(map(int, position[:-1].split(\",\"))))\n",
    "        except:\n",
    "            positions.append([])\n",
    "    return positions\n",
    "\n",
    "# Extract subsequence from GRCh37 file\n",
    "def ExtractSeq(chr_index,pos,length):\n",
    "    pos = pos+floor(pos/60)\n",
    "    hg19.seek(chrom_seek_index[chr_index,1]+pos-1)\n",
    "    return re.sub(r'\\n', '', hg19.read(length))\n",
    "\n",
    "# main blast function\n",
    "def Blast(query_seq):\n",
    "    i = 0\n",
    "    query_words = []\n",
    "    query_seq_length = len(query_seq)\n",
    "    words_length = query_seq_length-11+1\n",
    "    while i < words_length:\n",
    "        query_words.append(query_seq[i:i+11])\n",
    "        i += 1\n",
    "    words_positions = []\n",
    "    for word in query_words:\n",
    "        words_positions.append(GetWordPos(word))\n",
    "    for chr_index in range(24):\n",
    "        for word_index in range(words_length):\n",
    "            for pos in range(len(words_positions[word_index][chr_index])):\n",
    "                words_positions[word_index][chr_index][pos] += words_length - word_index - 1\n",
    "        \n",
    "        words_positions_corrects = []\n",
    "        for word_index in range(words_length):\n",
    "            words_positions_corrects += words_positions[word_index][chr_index]\n",
    "        \n",
    "        words_positions_corrects_count = Counter(words_positions_corrects)\n",
    "        finded_postions = []\n",
    "        for count_ in words_positions_corrects_count:\n",
    "            # we can select the bigger threshold of words_positions_corrects_count[count_] just \n",
    "            # like we select the highly similar sequence in NCBI BLAST\n",
    "            if words_positions_corrects_count[count_] > 5:\n",
    "                finded_postions.append(count_)\n",
    "        if finded_postions:\n",
    "            for finded_postion in finded_postions:\n",
    "                candidate_seq_pos = finded_postion - query_seq_length + 11 - 5\n",
    "                candidate_seq_length = query_seq_length + 11\n",
    "                candidate_sequence = ExtractSeq(chr_index,candidate_seq_pos,candidate_seq_length)\n",
    "                i_start_indexs = []\n",
    "                for i_start in range(15):\n",
    "                    _,_,score = SMalignment(candidate_sequence[i_start:],query_seq)\n",
    "                    i_start_indexs.append(score)\n",
    "                i_start = np.array(i_start_indexs).argmax()\n",
    "                i_end_indexs = []\n",
    "                for i_end in range(1,16):\n",
    "                    _,_,score = SMalignment(candidate_sequence[:-i_end],query_seq)\n",
    "                    i_end_indexs.append(score)\n",
    "                i_end = np.array(i_end_indexs).argmax()+1\n",
    "                candidate_sequence = candidate_sequence[i_start:-i_end]\n",
    "                align_seq1,align_seq2,align_score = SMalignment(candidate_sequence,query_seq)\n",
    "                if align_score>0.7:\n",
    "                    print(\"find in chromosome \"+chr_names[chr_index]+\": \"+str(candidate_seq_pos+i_start)+' ---> '+str(candidate_seq_pos+i_start+len(candidate_sequence)-1)+\", align score: \"+str(align_score))\n",
    "                    Display(align_seq1, align_seq2)\n",
    "    return None\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    chr_names = np.load('/home/jxiaoae/class/blast/GRCh37_chr_names.npy')\n",
    "    chrom_seek_index = np.load('/home/jxiaoae/class/blast/GRCh37_chrom_seek_index.npy')\n",
    "    hg19 = open(\"/home/share/GRCh37/human_g1k_v37.fasta\")\n",
    "    query_sequence = 'GTATCGGAACTTCCAACTTGTAGGCAAAATAGATATGCTTCATATTCTTAAAAACCACAAGAAA'\n",
    "    Blast(query_sequence)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
